aiida/engine/transports.py

```diff
@@ -23,11 +23,48 @@
 class TransportRequest:
     """ Information kept about request for a transport object """
 
-    def __init__(self):
+    def __init__(self, authinfo: AuthInfo):
         super().__init__()
         self.future: asyncio.Future = asyncio.Future()
         self.count = 0
 
+        self._callback_handle: Optional[asyncio.TimerHandle] = None
+        self._authinfo = authinfo
+        self._transport = authinfo.get_transport()
+
+    def do_open(self):
+        """ Actually open the transport
+
+        Sets self.future to the open transport, if successful, or to an exception.
+        """
+        if self.count > 0:
+
+            # The user still wants the transport so open it
+            _LOGGER.debug('Transport request opening transport for %s', self._authinfo)
+            try:
+                self._transport.open()
+            except Exception as exception:  # pylint: disable=broad-except
+                _LOGGER.error('exception occurred while trying to open transport:\n %s', exception)
+                self.future.set_exception(exception)
+
+                # Cleanup of the stale TransportRequest with the excepted transport future
+                # self._transport_requests.pop(authinfo.id, None)
```

Comment:
this is what would have happened in the previous codebase.

does this cleanup really need to happen here or is it ok to do it at the end?
I guess all the other context managers waiting for this transport will anyhow get the excepted one.

on the other hand it's true - if this is cleaned up here already, follow-up transport requests can already get a fresh one

it just makes the code more difficult to reason about